src/net.c

```diff
@@ -1535,12 +1535,58 @@ handle_request(coap_context_t *context, coap_queue_t *node) {
   assert(response == NULL);
 }
 
+static int
+coap_remove_separate_from_queue(coap_context_t *context, const coap_address_t *dst,
+			 const unsigned char *token, size_t token_length, coap_queue_t **sent) {
+  coap_queue_t *p, *q;
+  if (!context->sendqueue)
+    return 0;
+
+  if (coap_address_equals(dst, &context->sendqueue->remote) &&
+	    token_match(token, token_length,
+		              context->sendqueue->pdu->hdr->token,
+		              context->sendqueue->pdu->hdr->token_length) &&
+      context->sendqueue->retransmit_cnt > COAP_DEFAULT_MAX_RETRANSMIT) {
+    *sent = context->sendqueue;
+    context->sendqueue = context->sendqueue->next;
+    return 1;
+  }
+
+  p = context->sendqueue;
+  q = p->next;
+
+  /* when q is not NULL, it does not match (dst, token), so we can skip it */
+  while (q) {
+    if (coap_address_equals(dst, &context->sendqueue->remote) &&
+        token_match(token, token_length,
+                    context->sendqueue->pdu->hdr->token,
+                    context->sendqueue->pdu->hdr->token_length) &&
+        context->sendqueue->retransmit_cnt > COAP_DEFAULT_MAX_RETRANSMIT) {
+      p->next = q->next;
+      q->next = NULL;
+      *sent = q;
+      return 1;
+    } else {
+      p = q;
+      q = q->next;
+    }
+  }
+  return 0;
+}
+
 static inline void
 handle_response(coap_context_t *context, 
 		coap_queue_t *sent, coap_queue_t *rcvd) {
 
   coap_send_ack(context, &rcvd->local_if, &rcvd->remote, rcvd->pdu);
-  
+
+  /* Find separate response's request */
+  if (sent == NULL) {
```

Comment:
It took me some time to understand why this will actually work---can you include a comment that explains why in this case, the sendqueue may contain the sent request despite `sent` being `NULL`?